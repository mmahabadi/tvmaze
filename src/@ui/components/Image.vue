<script setup lang="ts">
import { defineProps } from 'vue'

defineProps<{ src?: string; alt?: string }>()
</script>
<template>
  <img v-if="src" :src="src" :alt="alt" class="w-full h-full object-cover" />
  <img
    v-else
    class="w-full h-full object-cover"
    src="@/assets/image-placeholder.jpg"
    alt="image-placeholder"
  />
</template>
