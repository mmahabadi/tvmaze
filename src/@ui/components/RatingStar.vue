<script setup lang="ts">
import { computed, defineProps } from 'vue'
import StarIconOutline from './icons/StarIconOutline.vue'
import StarIcon from './icons/StarIcon.vue'

const props = defineProps<{ rating: number }>()
const filledStars = computed(() => Math.floor(props.rating / 2))
</script>
<template>
  <div class="flex gap-2 text-yellow-500">
    <!-- Render filled stars -->
    <span v-for="index in filledStars" :key="`filled-star-${index}`">
      <StarIcon />
    </span>

    <!-- Render empty stars -->
    <span v-for="index in 5 - filledStars" :key="`empty-star-${index}`">
      <StarIconOutline />
    </span>
  </div>
</template>
