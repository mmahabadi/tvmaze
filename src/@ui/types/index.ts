export type SwiperOptions = {
  centeredSlides: boolean
  slidesPerView: number
  spaceBetween: number
  navigation: boolean
  loop: boolean
  scrollbar: { draggable: boolean }
  speed: number
  lazy: boolean
  autoplay: {
    delay: number
    disableOnInteraction: boolean
  }
}
