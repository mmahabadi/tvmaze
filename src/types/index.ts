export type Rating = {
  average: number
}

export type Schedule = {
  time: string
  days: string[]
}

export type Image = {
  medium: string
  original: string
}

export type Movie = {
  id: number
  name: string
  type: string
  language: string
  genres: string[]
  status: string
  //   runtime: 60
  //   averageRuntime: 60
  //   premiered: '2013-06-24'
  ended: string //'2015-09-10'
  officialSite: string //'http://www.cbs.com/shows/under-the-dome/'
  schedule: Schedule
  rating: Rating
  //   weight: 99
  //   network: {
  //     id: 2
  //     name: 'CBS'
  //     country: {
  //       name: 'United States'
  //       code: 'US'
  //       timezone: 'America/New_York'
  //     }
  //     officialSite: 'https://www.cbs.com/'
  //   }
  //   webChannel: null
  //   dvdCountry: null
  //   externals: {
  //     tvrage: 25988
  //     thetvdb: 264492
  //     imdb: 'tt1553656'
  //   }
  image: Image
  summary: string
  //   updated: 1704794065
  //   _links: {
  //     self: {
  //       href: 'https://api.tvmaze.com/shows/1'
  //     }
  //     previousepisode: {
  //       href: 'https://api.tvmaze.com/episodes/185054'
  //       name: 'The Enemy Within'
  //     }
  //   }
  _embedded: Embedded
}

export type Embedded = {
  cast: Cast[]
}

export type Cast = {
  person: Person
  character: Character
}

export type Person = {
  id: number
  url: string
  name: string
  country: Country
  birthday: string
  deathday: string | null
  image: Image
  gender: string
}

export type Country = {
  name: string
  code: string
  timezone: string
}

export type Character = {
  id: number
  url: string
  name: string
  image: Image
}

export type SlideItem = Pick<Movie, 'id' | 'name'> & { rating: number; image: string }
