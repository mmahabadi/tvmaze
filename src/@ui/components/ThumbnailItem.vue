<script setup lang="ts">
import { SlideItem } from '../types'
import RatingStar from './RatingStar.vue'
import PlayIcon from './icons/PlayIcon.vue'

defineProps<{ item: SlideItem }>()
</script>
<template>
  <div
    v-if="!!item"
    class="p-4 h-rate border border-border cursor-pointer bg-dry rounded-lg overflow-hidden"
  >
    <img
      :src="item.image"
      alt="About Time of the Dead"
      class="w-full h-full object-cover rounded-lg"
    />
    <div
      class="px-4 gap-6 text-center absolute hover:bg-black hover:bg-opacity-70 top-0 left-0 right-0 bottom-0 display-none flex flex-col justify-center items-center transition duration-300 ease-in-out opacity-0 hover:opacity-100 hover:display-block"
    >
      <button
        class="w-12 h-12 flex justify-center items-center transitions hover:bg-subMain rounded-full bg-white bg-opacity-30 text-white"
      >
        <PlayIcon />
      </button>
      <a class="font-semibold text-xl trancuted line-clamp-2" :href="item.name">{{ item.name }}</a>
      <RatingStar :rating="item.rating" />
    </div>
  </div>
</template>
