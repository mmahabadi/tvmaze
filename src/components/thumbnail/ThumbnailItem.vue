<script setup lang="ts">
import { RouterLink } from 'vue-router'
import { type SlideItem } from '@/types'
import RatingStar from '@/@ui/components/RatingStar.vue'
import IconPlay from '../icons/IconPlay.vue'
import Image from '@/@ui/components/Image.vue'

defineProps<{ item: SlideItem }>()
</script>
<template>
  <div v-if="!!item" class="container">
    <RouterLink :to="`/details/${item.id}`">
      <Image class="t-image" :src="item.image" :alt="item.name" />
      <div class="item">
        <button class="item-button">
          <IconPlay />
        </button>
        <a class="title" :href="item.name">{{ item.name }}</a>
        <RatingStar :rating="item.rating" />
      </div>
    </RouterLink>
  </div>
</template>
<style scoped>
.container {
  @apply p-4 border cursor-pointer rounded-lg overflow-hidden;

  RouterLink {
    @apply hover:text-teal-500 transition text-white flex justify-center items-center;
  }
  .t-image {
    @apply w-full h-full min-h-44 object-cover rounded-lg;
  }
  .item {
    @apply px-4 gap-6 text-center absolute hover:bg-black hover:bg-opacity-70 top-0 left-0 right-0 bottom-0 flex flex-col justify-center items-center transition duration-300 ease-in-out opacity-0 hover:opacity-100;

    .item-button {
      @apply w-12 h-12 flex justify-center items-center transition hover:bg-teal-500 rounded-full bg-white bg-opacity-30 text-white;
    }
    .title {
      @apply font-semibold w-full text-xl truncate overflow-hidden line-clamp-2;
    }
  }
}
</style>
