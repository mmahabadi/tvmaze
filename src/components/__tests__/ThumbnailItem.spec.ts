import { describe, it, expect, beforeEach } from 'vitest'
import { mount, RouterLinkStub, VueWrapper } from '@vue/test-utils'
import ThumbnailItem from '../thumbnail/ThumbnailItem.vue'
import { type SlideItem } from '@/types'
import RatingStar from '@/@ui/components/RatingStar.vue'
import IconPlay from '@/components/icons/IconPlay.vue'
import Image from '@/@ui/components/Image.vue'

describe('ThumbnailItem', () => {
  let wrapper: VueWrapper<unknown>
  const props = {
    item: {
      id: 1,
      name: 'Inception',
      image: 'Inception.jpg',
      rating: 8.5
    } as SlideItem
  }

  beforeEach(() => {
    wrapper = mount(ThumbnailItem, {
      props,
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
          RatingStar,
          IconPlay,
          Image
        }
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
