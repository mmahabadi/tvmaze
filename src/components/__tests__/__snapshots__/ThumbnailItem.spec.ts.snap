// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThumbnailItem > renders correctly 1`] = `
"<div data-v-8b7c7139="" class="container"><a data-v-8b7c7139=""><img data-v-8b7c7139="" src="Inception.jpg" alt="Inception" class="w-full h-full object-cover t-image">
    <div data-v-8b7c7139="" class="item"><button data-v-8b7c7139="" class="item-button"><svg data-v-8b7c7139="" class="svg-icon" style="width: 1em; height: 1em; vertical-align: middle; fill: currentColor; overflow: hidden;" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
          <path d="M852.727563 392.447107C956.997809 458.473635 956.941389 565.559517 852.727563 631.55032L281.888889 993.019655C177.618644 1059.046186 93.090909 1016.054114 93.090909 897.137364L93.090909 126.860063C93.090909 7.879206 177.675064-35.013033 281.888889 30.977769L852.727563 392.447107 852.727563 392.447107Z"></path>
        </svg></button><a data-v-8b7c7139="" class="title" href="Inception">Inception</a>
      <div data-v-8b7c7139="" class="flex gap-2 text-yellow-500">
        <!-- Render filled stars --><span><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 576 512" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"></path></svg></span><span><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 576 512" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"></path></svg></span><span><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 576 512" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"></path></svg></span><span><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 576 512" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"></path></svg></span><!-- Render empty stars --><span><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 576 512" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"></path></svg></span>
      </div>
    </div>
  </a></div>"
`;
